preprocess/foldfilter_main.cc

```diff
@@ -0,0 +1,320 @@
+	vector<int32_t> pos_delimiters(options.delimiters.size(), 0);
+
+	// Position of the first delimiter we encountered up to pos. Reset
+	// to pos + next char if it's not a delimiter.
+	int32_t pos_first_delimiter = 0;
+
+	while (pos < length) {
+		UChar32 character;
+
+		U8_NEXT(line.data(), pos, length, character);
+		
+		if (character < 0)
+			throw utf8::NotUTF8Exception(line);
+
+		size_t delimiter_idx = find_delimiter(options.delimiters, character);
+
+		if (delimiter_idx != not_found) {
+			// Store pos_first_delimiter instead of pos because when we have
+			// consecutive delimiters we want to chop em all off, even when
+			// our ideal delimiter is somewhere in the middle.
+			pos_delimiters[delimiter_idx] = pos_first_delimiter;
+		} else {
+			// Maybe the next char is a delimiter? pos is pointing to the next
+			// one right now, U8_NEXT incremented it.
+			pos_first_delimiter = pos;
+		}
+
+		// Do we need to introduce a break? If not, move to next character
+		if (pos - pos_last_cut < static_cast<int32_t>(options.column_width))
+			continue;
+
+		// Last resort if we didn't break on a delimiter: just chop where we are
+		int32_t pos_cut = pos;
+
+		// Find a more ideal break point by looking back for a delimiter
+		for (int32_t const &pos_delimiter : pos_delimiters) {
+			if (pos_delimiter > pos_last_cut) {
+				pos_cut = pos_delimiter;
+				break;
+			}
+		}
+
+		// Assume we cut without delimiters (i.e. the last resort scenario)
+		int32_t pos_cut_end = pos_cut;
+
+		// Peek ahead to were after the cut we encounter our first not-a-delimiter
+		// because that's the point were we resume.
+		for (int32_t pos_next = pos_cut_end; pos_cut_end < length; pos_cut_end = pos_next) {
+			// When we're not skipping delimiters, don't send more bytes than
+			// column_width in total a single line, even though we try to keep
+			// the delimiters together.
+			if (options.keep_delimiters && pos_cut_end - pos_last_cut >= static_cast<int32_t>(options.column_width))
+				break;
+
+			U8_NEXT(line.data(), pos_next, length, character);
+
+			if (character < 0)
+				throw utf8::NotUTF8Exception(line);
+
+			// First character after pos_cut is probably a delimiter, unless
+			// we did a hard stop in the middle of a word, and we're not keeping
+			// the delimiters.
+			if (find_delimiter(options.delimiters, character) == not_found)
+				break;
+		}
+
+		if (options.keep_delimiters) {
+			out_lines.push_back(line.substr(pos_last_cut, pos_cut_end - pos_last_cut));
+			out_delimiters.emplace_back("");
+		} else {
+			out_lines.push_back(line.substr(pos_last_cut, pos_cut - pos_last_cut));
+			out_delimiters.emplace_back(line.substr(pos_cut, pos_cut_end - pos_cut).data(), pos_cut_end - pos_cut);
+		}
+
+		pos_last_cut = pos_cut_end;
+		pos = pos_cut_end;
+	}
+
+	// Push out any trailing bits. Or the empty bit.
+	if (pos_last_cut < pos || pos == 0) {
+		out_lines.push_back(line.substr(pos_last_cut, pos - pos_last_cut));
```

Comment:
I'm not sure what you're referring to. `line` is a StringPiece, this if-statement is the part where there is no need to add another cut to meet the line length limit but we need to do something with the remainder (if there was any input to begin with)

Reply:
Sorry you're right I tagged the wrong line.  Is the delimiter always a UChar32?  Is `std::string` overkill for storing that?

Reply:
It's always a substring, zero or more characters from line itself. If I cut in the middle of a word or if delimiters are kept in the line, delimiters is always empty strings. If I do take out the delimiters, then it can be a longer string as it treats all consecutive delimiters as a single giant delimiter.

I'm not sure why I used `deque<StringPiece>` for `out_lines` but `deque<string>` for `out_delimiters`… Ah! Because `out_lines` gets put into the child process stdin and then I don't care about it anymore. `out_delimiters` goes into the queue, and outlives `line` which wouldn't work if they originate from `line.substr()`. So I needed to give them their own memory. `std::string` was the easiest trick for that.